---
title: "A lncRNA signature for metastasis prediction of LA-NPC"
output: html_document
editor_options: 
  chunk_output_type: console
---

This a R markdown file contains codes that reproduce the main results of our study. The required data format and expected outputs can be found in the corresponding html file.  

## Differential expression analysis
Differential expression analysis of microarray data was performed with GeneSpring GX v12.1 software (Agilent Technologies) and 149 differentially expressed (DE) lncRNAs were identified. Model development and validation, and the function implication analyses were performed using the following codes.

Load packages
```{r,message = FALSE,warning = FALSE}
library(dplyr)
library(ggplot2)
library(pROC)
library(tableone)
library(circlize)
library(ggplotify)
library(tibble)
library(tidyr)
library(survival)
library(survminer)
library(glmnet)
library(gridExtra)
library(cowplot)
library(Hmisc)
library(forestplot)
library(msigdbr)
library(clusterProfiler)
library(enrichplot)
library(sva)
library(pheatmap)
library(ggpubr)
library(limma)
library(stringr)
library(org.Hs.eg.db)
library(ComplexHeatmap)
library(GSVA)
library(MCPcounter)
library(ggunchained)
library(scales)

```


## Training
Input data
- TrainingData contains the expression of the 149 DE lncRNAs by qRT-PCR and clinical information of each patients in the training cohort.
```{r,message = FALSE,warning = FALSE}
TrainingData<-read.csv("Cohort-Training.csv",header = T,row.names = 1,check.names = F)
TrainingData[1:3,]
```

- DElnc contains the probe_id, gene symbol of the 149 DE lncRNAs and whether they were selected to model construction.
```{r}
DElnc <- read.csv("DE_lncRNA.csv",row.names = 1,header = T)
head(DElnc)
```

Univariate Cox analysis to identify lncRNAs related to DMFS
```{r,message = FALSE,warning = FALSE}
covariates1 <- colnames(TrainingData)[1:length(grep("ENST",colnames(TrainingData)))]
covariates2 <- c("AgeGroup","Sex","Tgroup","Ngroup","EBVDNAGroup")
covariates <- union(covariates1,covariates2)
univ_formulas1 <- sapply(covariates,function(x) as.formula(paste('Surv(DMFS, Metastasis)~', x)))
univ_models1 <- lapply(univ_formulas1, function(x){coxph(x, data = TrainingData)})
univ_results1 <- lapply(univ_models1,
                        function(x){ 
                          x <- summary(x)
                          wald.p.value<-sprintf("%.8f",x$waldtest["pvalue"])
                          wald.test<-signif(x$wald["test"], digits=3)
                          beta<-sprintf("%.2f",x$coef[1]);#coeficient beta
                          HR <-sprintf("%.2f",x$coef[2]);#exp(beta)
                          HR.confint.lower <- sprintf("%.2f",x$conf.int[,"lower .95"])
                          HR.confint.upper <- sprintf("%.2f",x$conf.int[,"upper .95"])
                          HR <- paste0(HR, " (", 
                                       HR.confint.lower, "-", HR.confint.upper, ")")
                          res<-c(beta, HR, wald.test, wald.p.value)
                          names(res)<-c("beta", "HR (95% CI for HR)", "wald.test", "wald.p.value")
                          return(res)
                        })
res <- t(as.data.frame(univ_results1, check.names = FALSE))
res <- as.data.frame(res)
exlist<-which(res$wald.p.value>=0.05)
TrainingData_origin<-TrainingData
TrainingData<-TrainingData_origin[,-exlist] 
#only lncRNAs related to DMFS were included in signature development
```

Model development using LASSO
```{r,message = FALSE,warning = FALSE}
x1<-as.matrix(TrainingData[,1:length(grep("ENST",colnames(TrainingData)))])         
y1<-cbind(time=TrainingData$DMFS,status=TrainingData$Metastasis)

set.seed(1)
cv.fit<-cv.glmnet(x1,y1,family="cox",nfolds=10,alpha=1,nlambda=1000)
modelcoef<-coef(cv.fit, s = cv.fit$lambda.min)
selected_lncs<-colnames(TrainingData)[which(modelcoef!=0)]
selected_lncs
DElnc[selected_lncs,"lncipedia_GeneSymbol"]
lncs_coef<-as.numeric(sprintf(" %.3f",modelcoef[which(modelcoef!=0)]))
lncs_coef
lncs_number<-length(selected_lncs)
riskscore <- 0
for (l in 1:lncs_number){
  coef <- lncs_coef[l]
  lnc <- TrainingData[,selected_lncs[l]]
  riskscore <- riskscore+coef*lnc
}
names(riskscore)<-"riskscore"
data_train<-cbind(TrainingData_origin,riskscore)

riskscore_ROC <- roc(data_train$Metastasis,data_train$riskscore,levels=c(control = 0, case = 1),direction="<")
cutoff <- riskscore_ROC$thresholds[which.max(riskscore_ROC$sensitivities+riskscore_ROC$specificities)]
data_train$lncRNA_signature<-cut(data_train$riskscore,breaks = c(-100,cutoff,100),labels = c(0,1))
```

Multivariate Cox analysis
```{r,message = FALSE,warning = FALSE,eval = FALSE}
summary(coxph(Surv(DMFS, Metastasis)~Ngroup+EBVDNAGroup+lncRNA_signature, data = data_train))
summary(coxph(Surv(DFS, Progression)~Ngroup+EBVDNAGroup+AgeGroup+lncRNA_signature, data = data_train))
summary(coxph(Surv(OS, Death)~AgeGroup+Ngroup+EBVDNAGroup+lncRNA_signature, data = data_train))

```

KM curves (Fig.2A-C)
```{r,message = FALSE,warning = FALSE}
modelplots <- list()
model_DMFS<- survfit(Surv(DMFS, Metastasis) ~ lncRNA_signature, data = data_train)  
modelplots[[1]]<-ggsurvplot(model_DMFS, data=data_train,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Distant Metastasis-free Survival",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,xlim=c(0,100),
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
model_DFS<- survfit(Surv(DFS, Progression) ~ lncRNA_signature, data = data_train)  
modelplots[[2]]<-ggsurvplot(model_DFS, data=data_train,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Disease-free Survival",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,xlim=c(0,100),
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
model_OS<- survfit(Surv(OS, Death) ~ lncRNA_signature, data = data_train)  
modelplots[[3]]<-ggsurvplot(model_OS, data=data_train,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Overall Survival",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,xlim=c(0,100),
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
arrange_ggsurvplots(modelplots, print = TRUE,  ncol = 3, nrow = 1)
```

ROC analysis (Fig.4A and 4D)
```{r,message = FALSE,warning = FALSE}
riskscore_ROC <- roc(data_train$Metastasis,data_train$riskscore,levels=c(control = 0, case = 1),direction="<")
EBV_ROC <- roc(data_train$Metastasis,data_train$EBVDNA,levels=c(control = 0, case = 1),direction="<")
Ngroup_ROC <- roc(data_train$Metastasis,data_train$Ngroup,levels=c(control = 0, case = 1),direction="<")
#riskscore_ROC$auc; ci(riskscore_ROC)
#Ngroup_ROC$auc
#EBV_ROC$auc
#a1 <- roc.test(riskscore_ROC,EBV_ROC);a1$p.value
#a2 <- roc.test(riskscore_ROC,Ngroup_ROC);a2$p.value
plot(Ngroup_ROC,col="blue",lwd=2)
plot(EBV_ROC,col="black",lwd=2,add=T)
plot(riskscore_ROC,col="red",lwd=2,add=T)

data_train$lncRNA_signature <- as.numeric(data_train$lncRNA_signature)-1
#summary(coxph(Surv(DMFS, Metastasis)~Ngroup+lncRNA_signature, data = data_train))
data_train$lnc_N_score <- data_train$Ngroup+2*data_train$lncRNA_signature
lncsig_ROC <- roc(data_train$Metastasis,data_train$lncRNA_signature,levels=c(control = 0, case = 1),direction="<")
lnc_N_ROC <- roc(data_train$Metastasis,data_train$lnc_N_score,levels=c(control = 0, case = 1),direction="<")
#lncsig_ROC$auc
#lnc_N_ROC$auc
#b1 <- roc.test(Ngroup_ROC,lnc_N_ROC);b1$p.value
#b2 <- roc.test(lncsig_ROC,lnc_N_ROC);b2$p.value
plot(Ngroup_ROC,col="blue",lwd=2)
plot(lnc_N_ROC,col="black",lwd=2,add=T)
plot(lncsig_ROC,col="red",lwd=2,add=T)
```


## Internal Validation
- ValidationData contains the expression of the 9 selected lncRNAs by qRT-PCR and clinical information of each patients in the internal validation cohort.
```{r,message = FALSE,warning = FALSE}
ValidationData<-read.csv("Cohort-Internal.csv",header = T,row.names = 1,check.names = F)
ValidationData[1:3,]
```

riskscore calsulation
```{r,message = FALSE,warning = FALSE}
riskscore_iv <- 0
for (l in 1:lncs_number){
  coef <- lncs_coef[l]
  lnc <- ValidationData[,selected_lncs[l]]
  riskscore_iv <- riskscore_iv+coef*lnc
}
ValidationData$riskscore<-riskscore_iv
ValidationData$lncRNA_signature<-cut(ValidationData$riskscore,breaks = c(-100,cutoff,100),labels = c(0,1))
data_invd <- ValidationData
```

Cox analysis
```{r,message = FALSE,warning = FALSE,eval = FALSE}
# Univariate Cox
# Variables include AgeGroup,Sex,Tgroup,Ngroup,EBVDNAGroup
summary(coxph(Surv(DMFS, Metastasis)~AgeGroup, data = data_invd))

# Multivariate Cox
summary(coxph(Surv(DMFS, Metastasis)~Ngroup+EBVDNAGroup+lncRNA_signature, data = data_invd))
summary(coxph(Surv(DFS, Progression)~Sex+Ngroup+EBVDNAGroup+lncRNA_signature, data = data_invd))
summary(coxph(Surv(OS, Death)~Ngroup+EBVDNAGroup+Sex+lncRNA_signature, data = data_invd))
```

KM curves (Fig.2D-F)
```{r,message = FALSE,warning = FALSE}
modelplots <- list()
model_DMFS<- survfit(Surv(DMFS, Metastasis) ~ lncRNA_signature, data = data_invd)  
modelplots[[1]]<-ggsurvplot(model_DMFS, data=data_invd,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Distant Metastasis-free Survival",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,xlim=c(0,100),
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
model_DFS<- survfit(Surv(DFS, Progression) ~ lncRNA_signature, data = data_invd)  
modelplots[[2]]<-ggsurvplot(model_DFS, data=data_invd,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Disease-free Survival",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,xlim=c(0,100),
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
model_OS<- survfit(Surv(OS, Death) ~ lncRNA_signature, data = data_invd)  
modelplots[[3]]<-ggsurvplot(model_OS, data=data_invd,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Overall Survival",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,xlim=c(0,100),
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
arrange_ggsurvplots(modelplots, print = TRUE,  ncol = 3, nrow = 1)
```

ROC analysis (Fig.4B and 4E)
```{r,message = FALSE,warning = FALSE}
riskscore_ROC_iv <- roc(data_invd$Metastasis,data_invd$riskscore,levels=c(control = 0, case = 1),direction="<")
EBV_ROC_iv <- roc(data_invd$Metastasis,data_invd$EBVDNA,levels=c(control = 0, case = 1),direction="<")
Ngroup_ROC_iv <- roc(data_invd$Metastasis,data_invd$Ngroup,levels=c(control = 0, case = 1),direction="<")
#riskscore_ROC_iv$auc; ci(riskscore_ROC_iv)
#EBV_ROC_iv$auc
#Ngroup_ROC_iv$auc
#roc.test(riskscore_ROC_iv,EBV_ROC_iv)
#roc.test(riskscore_ROC_iv,Ngroup_ROC_iv)
plot(Ngroup_ROC_iv,col="blue",lwd=2)
plot(EBV_ROC_iv,col="black",lwd=2,add=T)
plot(riskscore_ROC_iv,col="red",lwd=2,add=T)

data_invd$lncRNA_signature <- as.numeric(data_invd$lncRNA_signature)-1
data_invd$lnc_N_score <- data_invd$Ngroup+2*data_invd$lncRNA_signature
lncsig_ROC_iv <- roc(data_invd$Metastasis,data_invd$lncRNA_signature,levels=c(control = 0, case = 1),direction="<")
lnc_N_ROC_iv <- roc(data_invd$Metastasis,data_invd$lnc_N_score,levels=c(control = 0, case = 1),direction="<")
#lncsig_ROC_iv$auc
#lnc_N_ROC_iv$auc
#roc.test(Ngroup_ROC_iv,lnc_N_ROC_iv)
#roc.test(lncsig_ROC_iv,lnc_N_ROC_iv)
plot(Ngroup_ROC_iv,col="blue",lwd=2)
plot(lnc_N_ROC_iv,col="black",lwd=2,add=T)
plot(lncsig_ROC_iv,col="red",lwd=2,add=T)
```


## External validation
- Guilindata contains the expression of the 9 selected lncRNAs by qRT-PCR and clinical information of each patients in the external validation cohort.
```{r,message = FALSE,warning = FALSE}
Guilindata<-read.csv("Cohort-Guilin.csv",header = T,row.names = 1,check.names = F)
Guilindata[1:3,]
```

riskscore calsulation
```{r,message = FALSE,warning = FALSE}
riskscore_ev <- 0
for (l in 1:lncs_number){
  coef <- lncs_coef[l]
  lnc <- Guilindata[,selected_lncs[l]]
  riskscore_ev <- riskscore_ev+coef*lnc
}
Guilindata$riskscore<-riskscore_ev
Guilindata$lncRNA_signature<-cut(Guilindata$riskscore,breaks = c(-100,cutoff,100),labels = c(0,1))
```

KM curves (Fig.2G-I)
```{r,message = FALSE,warning = FALSE}
modelplots <- list()
model_DMFS<- survfit(Surv(DMFS, Metastasis) ~ lncRNA_signature, data = Guilindata)  
modelplots[[1]]<-ggsurvplot(model_DMFS, data=Guilindata,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Distant Metastasis-free Surveval",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
model_DFS<- survfit(Surv(DFS, Progression) ~ lncRNA_signature, data = Guilindata)  
modelplots[[2]]<-ggsurvplot(model_DFS, data=Guilindata,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=12,
                            xlab="Months",break.time.by=12,ylab="Disease-free Surveval",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
model_OS<- survfit(Surv(OS, Death) ~ lncRNA_signature, data = Guilindata)  
modelplots[[3]]<-ggsurvplot(model_OS, data=Guilindata,palette = c("#5a72b5","#ea5a49"),size=0.5,
                            conf.int = F,pval = T,censor=F,pval.size=5.25,pval.coord=c(12,0.2),
                            legend.title="lncRNA signature",legend.labs=c("low risk","high risk"),
                            legend=c(0.7,0.2),font.legend=15,
                            xlab="Months",break.time.by=12,ylab="Overall Surveval",surv.scale="percent",font.x=12,font.y=12,axes.offset=F,font.tickslab=12,
                            risk.table = T,tables.height=0.25,risk.table.fontsize=5.25,tables.theme = theme_cleantable())
arrange_ggsurvplots(modelplots, print = TRUE,  ncol = 3, nrow = 1)
```

Cox analysis
```{r,message = FALSE,warning = FALSE,eval = FALSE}
# Univariate Cox
# Variables include AgeGroup,Sex,Tgroup,Ngroup
summary(coxph(Surv(DMFS, Metastasis)~AgeGroup, data = Guilindata))

# Multivariate Cox
summary(coxph(Surv(DMFS, Metastasis)~Ngroup+lncRNA_signature, data = Guilindata))
summary(coxph(Surv(DFS, Progression)~lncRNA_signature, data = Guilindata))
summary(coxph(Surv(OS, Death)~lncRNA_signature, data = Guilindata))
```

ROC analysis (Fig.4C and 4F)
```{r,message = FALSE,warning = FALSE}
riskscore_ROC_ev <- roc(Guilindata$Metastasis,Guilindata$riskscore,levels=c(control = 0, case = 1),direction="<")
Ngroup_ROC_ev <- roc(Guilindata$Metastasis,Guilindata$Ngroup,levels=c(control = 0, case = 1),direction="<")
#riskscore_ROC_ev$auc; ci(riskscore_ROC_ev)
#Ngroup_ROC_ev$auc
#roc.test(riskscore_ROC_ev,Ngroup_ROC_ev)
plot(Ngroup_ROC_ev,col="blue",lwd=2)
plot(riskscore_ROC_ev,col="red",lwd=2,add=T)

Guilindata$lncRNA_signature <- as.numeric(Guilindata$lncRNA_signature)-1
Guilindata$lnc_N_score <- Guilindata$Ngroup+2*Guilindata$lncRNA_signature
lncsig_ROC_ev <- roc(Guilindata$Metastasis,Guilindata$lncRNA_signature,levels=c(control = 0, case = 1),direction="<")
lnc_N_ROC_ev <- roc(Guilindata$Metastasis,Guilindata$lnc_N_score,levels=c(control = 0, case = 1),direction="<")
#lncsig_ROC_ev$auc
#lnc_N_ROC_ev$auc
#roc.test(Ngroup_ROC_ev,lnc_N_ROC_ev)
#roc.test(lncsig_ROC_ev,lnc_N_ROC_ev)
plot(Ngroup_ROC_ev,col="blue",lwd=2)
plot(lnc_N_ROC_ev,col="black",lwd=2,add=T)
plot(lncsig_ROC_ev,col="red",lwd=2,add=T)
```


forest plot (Fig.3)\
- forestdata is the summary of univariate Cox analyses. Prepare it in a .csv file using excel.
```{r,message = FALSE,warning = FALSE}
forestdata<-read.csv("forest/forest_uni_DMFSplot.csv")
forestplot(as.matrix(forestdata[,1:5]), 
           forestdata$mean,
           forestdata$lower,
           forestdata$upper,
           is.summary=c(T,F,
                        T,F,F,F,T,F,
                        T,F,F,F,T,F,
                        T,F,F,F,T,F,
                        T,F,F,F,T,F,
                        T,F,F,T,F,
                        T,F,F,F,T),
           graph.pos= 4, graphwidth = unit(50,"mm"),
           clip=c(0.25,16),xlog = T, xticks = c(0.25,0.5,1.0,2.0,4.0,8.0,16.0),
           boxsize=0.3,txt_gp = fpTxtGp(cex=1.25),line.margin=unit(5,"mm"), lineheight = unit(10,'mm'),
           hrzl_lines = list("2" = gpar(lty=1)),
           col=fpColors(box="blue",summary="blue",lines = 'black'))
```

## Function implication
Input data\
- Normdata: Normalized data with probename\
- Column 1: probename\
- Column 2 to Column 37 contain data from 18 Healthy controls and 18 matched LA-NPC\
- Column 38 to Column 57 contain data from 10 LA-NPC without metastasis and 10 matched LA-NPC with metastasis
```{r,message = FALSE,warning = FALSE}
Normdata<-read.csv("Matrix.csv",header=T)
Normdata[1:5,1:5]
```

- Traits: Information of the 56 samples subjected to microarray profiling,including sample ID, group, batch and risk stratification(only for LA-NPC samples) according to lncRNA signature
```{r,message = FALSE,warning = FALSE}
Traits <- read.csv("Array_Traits.csv",header = T)
Traits[1:3,]
```

- anno: Probe annotation
```{r,message = FALSE,warning = FALSE}
anno <- read.csv("Probe_anno.csv",header = T)
anno[1:3,]
```

Data Organization
```{r,message = FALSE,warning = FALSE}
datamat <- Normdata[,20:57] #Column 20 to Column 57 contain data from LA-NPC
model <- model.matrix(~1,data = data.frame(Traits$Group[19:56]))
combatdata <- ComBat(dat = as.matrix(datamat), batch = Traits$Batch[19:56], mod = model)#remove batch effect
#identical(Normdata$ProbeName,anno$ProbeName)

##mRNA data
rowmRNA <- which(anno$biotype=="protein_coding"&anno$quality_filter=="pass")
mRNA <-combatdata[rowmRNA,]
mRNA <- data.frame(GeneSymbol=anno$GeneSymbol[rowmRNA],mRNA)
mRNA<-mRNA%>%group_by(GeneSymbol)%>%summarise_all(median)%>%as.data.frame()
mRNA <- column_to_rownames(mRNA,"GeneSymbol")

##selected-lncRNA data
rowlnc <- which(anno$Trans_ID %in% selected_lncs) #no duplicates
#anno[rowlnc,"quality_filter"] #all pass quality control
lnc<-combatdata[rowlnc,]
rownames(lnc) <- anno$Trans_ID[rowlnc]
```

lncRNA function implication - hallmarks pathways
```{r,message = FALSE,warning = FALSE}
dd <- data.frame(msigdbr(species = "Homo sapiens"))
hallmarks <- dd %>% filter(gs_cat == "H") %>% dplyr::select(gs_name,gene_symbol)

hallmark_df<-data.frame()
for (j in selected_lncs){
  y <- as.numeric(lnc[j,])
  rownames <- rownames(mRNA)
  cor_data_df <- data.frame(rownames)
  #correlation analysis
  for (i in 1:length(rownames)){
    test <- cor.test(as.numeric(mRNA[i,]),y,method="spearman")
    cor_data_df[i,2] <- test$estimate
    cor_data_df[i,3] <- test$p.value
  }
  names(cor_data_df) <- c("Genesymbol","correlation","pvalue")
  
  #prepare data for GSEA
  geneList <- cor_data_df$correlation
  names(geneList) = cor_data_df$Genesymbol
  geneList = sort(geneList, decreasing = TRUE)
  
  #GSEA
  set.seed(1)
  ResGSEA_hallmark <- GSEA(geneList,TERM2GENE = hallmarks,pvalueCutoff = 1,nPermSimple = 10000) 
  
  tmp<-ResGSEA_hallmark@result
  tmp$lnc<-j
  hallmark_df<-rbind(hallmark_df,tmp)
}
hallmark_sig <- filter(hallmark_df,pvalue < 0.05,qvalues < 0.25)
hallmark_sig <- data.frame(table(hallmark_sig$ID))
```

lncRNA function implication - Immport (Fig.5B)
```{r,message = FALSE,warning = FALSE}
Immport_list <- read.csv("Immport_GeneList.csv",header = T) #Dowloaded from Immport website (https://www.immport.org/home)
Immport_list <- Immport_list[,c(6,1)]

Immport_df<-data.frame()
for (j in selected_lncs){
  y <- as.numeric(lnc[j,])
  rownames <- rownames(mRNA)
  cor_data_df <- data.frame(rownames)
  #correlation analysis
  for (i in 1:length(rownames)){
    test <- cor.test(as.numeric(mRNA[i,]),y,method="spearman")
    cor_data_df[i,2] <- test$estimate
    cor_data_df[i,3] <- test$p.value
  }
  names(cor_data_df) <- c("Genesymbol","correlation","pvalue")
  
  #prepare data for GSEA
  geneList <- cor_data_df$correlation
  names(geneList) = cor_data_df$Genesymbol
  geneList = sort(geneList, decreasing = TRUE)
  
  #GSEA
  set.seed(1)
  ResGSEA_Immport <- GSEA(geneList,TERM2GENE = Immport_list,minGSSize = 3,pvalueCutoff = 1,nPermSimple = 10000)
  
  tmp<-ResGSEA_Immport@result
  tmp$lnc<-j
  Immport_df<-rbind(Immport_df,tmp)
}
# bubble plot
Immport_df<-within (Immport_df, {
  group <- NA 
  group [pvalue < 0.05 & qvalues < 0.25]<- 1
  group [pvalue > 0.05 | qvalues > 0.25]<- 0})
ggplot(Immport_df,aes(x=lnc,y=ID))+
  geom_point(aes(size=abs(NES),
                 color=group))+
  theme_bw()+
  theme(panel.grid = element_blank(),
        axis.text.x=element_text(angle=45,hjust = 1,vjust=0.5))+
  scale_color_gradient(low="lightgrey",high="#E82B2B")+
  labs(x=NULL,y=NULL)+
  guides(size=guide_legend(order=3))
```

pathways enriched in high-risk and low-risk group (Fig.5C)
```{r,message = FALSE,warning = FALSE}
mRNA_dat <- t(mRNA)
#identical(rownames(mRNA_dat),Traits$ID[19:56]) #TRUE
mRNA_dat <- data.frame(cbind(mRNA_dat,lncRNA_signature=Traits$lncRNA_signature[19:56]))

Traits2 <- Traits[19:56,]
#identical(rownames(mRNA_dat),Traits2$ID) #TRUE
Traits2$lncRNA_signature <- as.factor(Traits2$lncRNA_signature)
difmod<-model.matrix(~0+lncRNA_signature, data = Traits2)
colnames(difmod) <- c("highrisk","lowrisk")
conmat<-makeContrasts(contrasts="lowrisk-highrisk",levels=difmod)
fit <- lmFit(mRNA, difmod)
fit1 <- contrasts.fit(fit,conmat)
fit2 <- eBayes(fit1)
Res_limma<-topTable(fit2,coef="lowrisk-highrisk",n=nrow(fit2),adjust="BH")
geneList <- Res_limma$logFC
names(geneList) = rownames(Res_limma)
geneList = sort(geneList, decreasing = TRUE)
set.seed(10)
ResGSEA <- GSEA(geneList,TERM2GENE = hallmarks,pvalueCutoff = 1,nPermSimple = 10000)
tmp1<-ResGSEA@result

##circus plot
dat <- tmp1[which(tmp1$pvalue<0.05),]
dat$ID <- gsub("HALLMARK_", "", dat$ID)
#dat$ID
dat$category <- c('Immune-related','Cancer-related','Immune-related','Other','Cancer-related',
                  'Cancer-related','Immune-related','Cancer-related','Immune-related','Other',
                  "Metabolism-related",'Other','Immune-related','Cancer-related','Cancer-related',
                  'Other','Immune-related',"Metabolism-related",'Cancer-related','Immune-related',
                  'Cancer-related','Other')
dat$gene_num.min <- 0
dat$gene_num.max <- dat$setSize
dat$log10Pvalue <- -log10(dat$pvalue)
dat$rich.factor=abs(dat$NES)
dat$Count <- round(as.numeric(substring(dat$leading_edge, 6, 7))/100*dat$setSize)
dat$up.regulated <- dat$Count
dat<-within (dat, {up.regulated [NES < 0]<- 0})
dat$down.regulated <- dat$Count
dat<-within (dat, {down.regulated [NES > 0]<- 0})
rownames(dat) <- dat$ID
dat <- dat[order(dat$category),]
dat[1:3,]

#outmost circle (Track1)
circos.par(gap.degree = 0.5, start.degree = 90)
plot_data <- dat[c('ID', 'gene_num.min', 'gene_num.max')]
color_assign <- c('Cancer-related' = "#5684B9", 'Immune-related' = "#B41F1B", 
                  'Metabolism-related' = "#E6AE64",'Other'="#BDDCA6")
ko_color <- color_assign[dat$category]

circos.genomicInitialize(plot_data, plotType = NULL, major.by = 1)
circos.track(
  ylim = c(0, 1), track.height = 0.05, bg.border = NA, bg.col = ko_color,
  panel.fun = function(x, y) {
    ylim = get.cell.meta.data('ycenter')#ylim、xlim
    xlim = get.cell.meta.data('xcenter')
    sector.name = get.cell.meta.data('sector.index')#sector.name
    circos.axis(h = 'top', labels.cex = 0.4, labels.niceFacing = FALSE)
    circos.text(xlim, ylim, sector.name, cex = 0.4, niceFacing = FALSE)
  } )

#Track2
plot_data <- dat[c('ID', 'gene_num.min', 'Count', 'log10Pvalue')]
label_data <- dat['Count']
p_max <- round(max(dat$'log10Pvalue')) + 1  
colorsChoice <- colorRampPalette(c('#FF906F', '#861D30'))
color_assign <- colorRamp2(breaks = 0:p_max, col = colorsChoice(p_max + 1))

circos.genomicTrackPlotRegion(
  plot_data, track.height = 0.08, bg.border = NA, stack = TRUE,
  panel.fun = function(region, value,...) {
    circos.genomicRect(region, value, col = color_assign(value[[1]]), border = NA,...)
    ylim = get.cell.meta.data('ycenter')  
    xlim = label_data[get.cell.meta.data('sector.index'),1] / 2
    sector.name = label_data[get.cell.meta.data('sector.index'),1]
    #circos.text(xlim, ylim, sector.name, cex = 0.4, niceFacing = FALSE)
  } )

#Track3
dat$all.regulated <- dat$up.regulated + dat$down.regulated
dat$up.proportion <- dat$up.regulated / dat$all.regulated
dat$down.proportion <- dat$down.regulated / dat$all.regulated

dat$up <- dat$up.proportion * dat$gene_num.max
plot_data_up <- dat[c('ID', 'gene_num.min', 'up')]
names(plot_data_up) <- c('ID', 'start', 'end')
plot_data_up$type <- 1 

dat$down <- dat$down.proportion * dat$gene_num.max + dat$up
plot_data_down <- dat[c('ID', 'up', 'down')]
names(plot_data_down) <- c('ID', 'start', 'end')
plot_data_down$type <- 2 

plot_data <- rbind(plot_data_up, plot_data_down)
label_data <- dat[c('up', 'down', 'up.regulated', 'down.regulated')]
color_assign <- colorRamp2(breaks = c(1, 2), col = c("#FF5B62", "#7B4FE9"))

circos.genomicTrackPlotRegion(
  plot_data, track.height = 0.08, bg.border = NA, stack = TRUE, 
  panel.fun = function(region, value, ...) {
    circos.genomicRect(region, value, col = color_assign(value[[1]]), border = NA, ...) 
    ylim = get.cell.meta.data('cell.bottom.radius') - 0.5 
    xlim = label_data[get.cell.meta.data('sector.index'),1] / 2
    sector.name = label_data[get.cell.meta.data('sector.index'),3]
    xlim = (label_data[get.cell.meta.data('sector.index'),2]+label_data[get.cell.meta.data('sector.index'),1]) / 2
    sector.name = label_data[get.cell.meta.data('sector.index'),4]
  } )

#Track4
plot_data <- dat[c('ID', 'gene_num.min', 'gene_num.max', 'rich.factor')] 
label_data <- dat['category']  
color_assign <- c('Cancer-related' = "#5684B9", 'Immune-related' = "#B41F1B", 
                  'Metabolism-related' = "#E6AE64",'Other'="#BDDCA6")
circos.genomicTrack(
  plot_data, ylim = c(0, 1), track.height = 0.3, bg.col = 'gray95', bg.border = NA,
  panel.fun = function(region, value, ...) {
    sector.name = get.cell.meta.data('sector.index')  #sector.name 
    circos.genomicRect(region, value*0.3, col = color_assign[label_data[sector.name,1]], border = NA, ytop.column = 1, ybottom = 0, ...) 
    circos.lines(c(0, max(region)), c(0.3,0.3), col = 'gray', lwd = 0.3) 
  } )

category_legend <- Legend(
  labels = c('Malignant property','Immune-related', 'Metabolism-related', 'Other'),
  type = 'points', pch = NA,
  background = c("#5684B9","#B41F1B","#E6AE64","#BDDCA6"),
  labels_gp = gpar(fontsize = 8), grid_height = unit(0.5, 'cm'), grid_width = unit(0.5, 'cm'))

updown_legend <- Legend(
  labels = c('Enriched in the low-risk group', 'Enriched in the low-high group'), 
  type = 'points', pch = NA, background = c('red', 'blue'), 
  labels_gp = gpar(fontsize = 8), grid_height = unit(0.5, 'cm'), grid_width = unit(0.5, 'cm'))

pvalue_legend <- Legend(
  col_fun = colorRamp2(round(seq(0, p_max, length.out = 6), 0), 
                       colorRampPalette(c('#FF906F', '#861D30'))(6)),
  legend_height = unit(3, 'cm'), labels_gp = gpar(fontsize = 8), 
  title_gp = gpar(fontsize = 9), title_position = 'topleft', title = '-Log10(Pvalue)')

lgd_list_vertical <- packLegend(category_legend, updown_legend, pvalue_legend)
grid.draw(lgd_list_vertical)

circos.clear()
```

heatmap of immune-related genes (Fig.5D)
```{r,message = FALSE,warning = FALSE}
heatdat <- mRNA[c("GZMA","GZMB","STAT1","CXCL3","CXCL9","CXCL13","CXCL14","IL2","HLA-A", "HLA-B", "HLA-DRA","HLA-DRB1","HLA-DRB4","HLA-DQB1","HLA-DPB1","PTPRC","MS4A1","CD3D","CD3G","ITGAX"),]
heatdat <- heatdat[,c("NPC_08","NPC_22","NPC_31","NPC_29","NPC_07","NPC_21",
                      "NPC_32","NPC_36","NPC_12","NPC_09","NPC_14","NPC_38",
                      "NPC_35","NPC_11","NPC_17","NPC_18","NPC_19","NPC_20",
                      "NPC_16","NPC_30","NPC_10","NPC_03","NPC_04","NPC_33",
                      "NPC_05","NPC_02","NPC_24","NPC_25","NPC_23","NPC_13",
                      "NPC_15","NPC_26","NPC_01","NPC_34","NPC_27","NPC_28",
                      "NPC_06","NPC_37")]
annotation_col1 <- data.frame(group=factor(Traits$lncRNA_signature[19:56]))
rownames(annotation_col1) = Traits$ID[19:56]
ann_colors1 = list(Type = c("0" = "#33A02CFF", "1" = "#FF7F00FF"))
pheatmap(as.matrix(heatdat), scale = "row", cluster_row = FALSE,cluster_col = FALSE,
         treeheight_row=0,treeheight_col=25,
         color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
         border_color=FALSE,fontsize=7,show_rownames = T,show_colnames = F,
         annotation_col = annotation_col1,annotation_colors = ann_colors1)

```

## Immune infiltration
MCPcounter (Fig.6A)
```{r,message = FALSE,warning = FALSE}
genes <- data.table::fread("Signatures/genes.txt",data.table = F)
probesets <- data.table::fread("Signatures/probesets.txt",data.table = F,header = F)
lncArray_MCPcounter_results<- MCPcounter.estimate(mRNA,
                                                  featuresType= "HUGO_symbols",
                                                  probesets=probesets,
                                                  genes=genes)
dd1 <- data.frame(t(lncArray_MCPcounter_results)) %>% rownames_to_column()
#identical(dd1$rowname,Traits$ID[19:56]) #TRUE
dd1$Group2 <- Traits$lncRNA_signature[19:56]
dd2 <- dd1 %>% pivot_longer(cols = 2:11,names_to = "celltype",values_to = "score")
dd2$Group2 <- factor(dd2$Group2,levels = c("low-risk","high-risk"))

colors = c("#5a72b5","#ea5a49")
su2 = group_by(dd2,Group2,celltype) %>% 
  summarise(s = sd(score),score = mean(score))
p = ggplot(dd2, aes(x = celltype, y = score,fill= Group2))+
  geom_split_violin(trim = FALSE,color = NA,adjust = 1.5)+
  guides(fill=guide_legend(title="group"))+
  scale_fill_manual(values = colors)+
  theme_bw()
p + geom_point(data = su2,aes(x = celltype, y = score),
               pch=19,position=position_dodge(0.3),size=2.5)+
  geom_errorbar(data = su2,
                mapping = aes(ymin = score-s, ymax = score+s),
                width=0, 
                position=position_dodge(0.3), 
                color="black",
                size=0.5)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1,vjust = 1, colour = "black"))+
  theme(panel.grid.major =element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(),axis.line = element_line(colour = "black"))+
  theme(strip.background = element_blank(),strip.text = element_text(size = 10))+
  stat_compare_means(aes(group = Group2), label = "p.format",label.y = 12, method="t.test")
```

Immune infiltration by IHC image analysis (Fig.6D)\
- IHCdata: Results of IHC image analysis
```{r,message = FALSE,warning = FALSE}
IHCdata <- read.csv("SYSUdata_IHC.csv",header = T,row.names = 1)
head(IHCdata)

common_col <- intersect(colnames(data_train),colnames(data_invd))
data_train2 <- data_train[,common_col]
data_invd2 <- data_invd[,common_col]
SYSUdata <- rbind(data_train2,data_invd2)

loc <- match(rownames(IHCdata),rownames(SYSUdata))
SYSU_IHCdata <- cbind(IHCdata,SYSUdata[loc,])
SYSU_IHCdata<-within (SYSU_IHCdata, {
  riskgroup <-NA 
  riskgroup [lncRNA_signature == 0]<-"low-risk"
  riskgroup [lncRNA_signature == 1]<-"high-risk"})

##boxplot
p1 <- ggboxplot(SYSU_IHCdata, x = "riskgroup", y = "Intratumor_CD20",
          color = "riskgroup",palette = c("#000091","#ea5a49"),
          add = c("jitter"),width = 0.3,
          bxp.errorbar=T,bxp.errorbar.width = 0.2)+
  theme(legend.position = "none")+
  scale_y_continuous(labels = scientific)+
  labs(x="",y="Intratumoral CD20+ B cell count(10e3/mm2)")+
  stat_compare_means(aes(group = riskgroup), label = "p.format", method="t.test")

p2 <- ggboxplot(SYSU_IHCdata, x = "riskgroup", y = "Stromal_CD20",
          color = "riskgroup",palette = c("#000091","#ea5a49"),
          add = c("jitter"),width = 0.3,
          bxp.errorbar=T,bxp.errorbar.width = 0.2)+
  theme(legend.position = "none")+
  scale_y_continuous(labels = scientific)+
  labs(x="",y="Stromal CD20+ B cell count(10e3/mm2)")+
  stat_compare_means(aes(group = riskgroup), label = "p.format", method="t.test")

p3 <- ggboxplot(SYSU_IHCdata, x = "riskgroup", y = "Intratumor_CD8",
          color = "riskgroup",palette = c("#000091","#ea5a49"),
          add = c("jitter"),width = 0.3,
          bxp.errorbar=T,bxp.errorbar.width = 0.2)+
  theme(legend.position = "none")+
  scale_y_continuous(labels = scientific)+
  labs(x="",y="Intratumoral CD8+ T cell count(10e3/mm2)")+
  stat_compare_means(aes(group = riskgroup), label = "p.format", method="t.test")

p4 <- ggboxplot(SYSU_IHCdata, x = "riskgroup", y = "Stromal_CD8",
          color = "riskgroup",palette = c("#000091","#ea5a49"),
          add = c("jitter"),width = 0.3,
          bxp.errorbar=T,bxp.errorbar.width = 0.2)+
  theme(legend.position = "none")+
  scale_y_continuous(labels = scientific)+
  labs(x="",y="Stromal CD8+ T cell count(10e3/mm2)")+
  stat_compare_means(aes(group = riskgroup), label = "p.format", method="t.test")

plot_grid(p1, p2, p3, p4, ncol = 4)
```

